### PR TITLE
RES-2942 Circular Buffer recipe for RDBI

### PR DESCRIPTION
**Pull request by michelle-gu**
Still testing

Considered pipelining then trimming the list for addAll, but ultimately decided to maintain the list at "maxSize" -- not sure which way is actually better for performance/how much storage is being used

Did not implement removeAll(collection), retainAll(collection), iterator(), toArray(), toArray(array), remove(object) -- are they necessary?

### CODE REVIEW

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+
+/**
+ * A circular buffer in redis. Head of the queue is the element at index 0 in redis.
+ */
+public class RedisCircularBuffer<ValueType> implements Queue<ValueType> {
```
Comment (2bitoperations) [inline rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java]: consider also that your use of this inside our product is going to probably need a way to peek the whole list in redis.

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+            Set<ValueType> currentSet = new HashSet<ValueType>();
+            final int size = handle.jedis().llen(key).intValue();
+            final Set<Response<String>> responses = Sets.newHashSetWithExpectedSize(size);
+
+            for (int i = 0; i < size; i++) {
+                responses.add(pipeline.lindex(key, i));
+            }
+            pipeline.sync();
+
+            for(final Response<String> res : responses) {
```
Comment (KhaiNguyen) [inline rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java]: style thing 
```suggestion
            for (final Response<String> res : responses) {
```

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+    public List<ValueType> peekAll() {
+        try (final Handle handle = rdbi.open()) {
+            final Pipeline pipeline = handle.jedis().pipelined();
+            List<ValueType> currentList = new ArrayList<ValueType>();
+            final int size = handle.jedis().llen(key).intValue();
+            final List<Response<String>> responses = Lists.newArrayListWithCapacity(size);
+
+            for (int i = 0; i < size; i++) {
+                responses.add(pipeline.lindex(key, i));
```
Comment (2bitoperations) [inline rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java]: i think it is probably better to just get the whole thing as lrange(...) in one shot, and probably best to not bother fetching the size
(again, doing it as a two-step you are leaving yourself open to concurrency problems. imagine that while thread A is in the middle of iterating through a peekAll, thread B starts rapidly popping members out.)

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+            }
+            pipeline.sync();
+
+            for(final Response<String> res : responses) {
```
Comment (2bitoperations) [inline rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java]: also reformat the file

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    @Override
+    public boolean containsAll(Collection<?> c) {
+        try (final Handle handle = rdbi.open()) {
+            final Pipeline pipeline = handle.jedis().pipelined();
+            Set<ValueType> currentSet = new HashSet<ValueType>();
+            final int size = handle.jedis().llen(key).intValue();
+            final Set<Response<String>> responses = Sets.newHashSetWithExpectedSize(size);
+
+            for (int i = 0; i < size; i++) {
+                responses.add(pipeline.lindex(key, i));
```
Comment (2bitoperations) [inline rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java]: same as peekAll, i think you should get all in one shot to redis, given how we plan to use this.

further, consider the fact that you could probably use peekAll as a replacement for most of what is currently in this method.

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        return currentList.containsAll(c);
+    }
+
+    @Override
+    public boolean addAll(Collection<? extends ValueType> toAdd) {
+        for (final ValueType value : toAdd) {
+            final String valueAsString = serializationHelper.encode(value);
+            try (Handle handle = rdbi.open()) {
+                String valueStr = serializationHelper.encode(value);
+                final int newSize = handle.attach(RedisCircularBufferDAO.class).add(key, valueStr, maxSize);
```
Comment (2bitoperations) [inline rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java]: I think you can do this as a pipeline.

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBufferDAO.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+import com.lithium.dbi.rdbi.BindArg;
+import com.lithium.dbi.rdbi.BindKey;
+import com.lithium.dbi.rdbi.Query;
+
+public interface RedisCircularBufferDAO {
+
+    @Query(
+            "local size = redis.call('RPUSH', $key$, $valueToAdd$)\n" +
+                    "if size > tonumber($maxSize$) then\n" +
```
Comment (2bitoperations) [inline rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBufferDAO.java]: what if something has gone wrong and the current size is way larger than max size?

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java (hunk 6 above)
Comment (2bitoperations) [reply rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/queue/RedisCircularBuffer.java]: can't do as a pipeline (yet)

**Pull request merged**